packages/runtime-core/src/componentOptions.ts

```diff
@@ -954,23 +839,197 @@ export function resolveMergedOptions(
+    if (!fromValue) return toValue
+    if (!toValue) return fromValue
+    return extend(Object.create(null), toValue, fromValue)
+  },
+  provide(
+    toValue: Data | Function | undefined,
+    fromValue: Data | Function | undefined,
+    publicThis: ComponentPublicInstance
+  ) {
+    return function mergeProvideFn() {
+      return extend(
+        Object.create(null),
+        isFunction(toValue) ? toValue.call(publicThis) : toValue,
+        isFunction(fromValue) ? fromValue.call(publicThis) : fromValue
+      )
+    }
+  },
+  render(
+    toValue: InternalRenderFunction | undefined,
+    fromValue: InternalRenderFunction | undefined
+  ) {
+    return fromValue || toValue
+  }
+}
+;([
+  'beforeCreate',
+  'created',
+  'beforeMount',
+  'mounted',
+  'beforeUpdate',
+  'updated',
+  'activated',
+  'deactivated',
+  'beforeDestroy',
+  'beforeUnmount',
+  'destroyed',
+  'unmounted',
+  'renderTracked',
+  'renderTriggered',
+  'errorCaptured'
+] as const).forEach(hookName => {
+  defaultMergeStrategies[hookName] = function mergeHook(
+    toValue?: Function | Function[],
+    fromValue?: Function
+  ) {
+    if (!fromValue) return toValue
+    toValue = toValue ? (isArray(toValue) ? toValue : [toValue]) : undefined
+    return toValue ? toValue.concat(fromValue) : fromValue
+  }
+})
+;(['components', 'directives'] as const).forEach(assetName => {
+  defaultMergeStrategies[assetName] = function mergeAssets(
+    toValue: Record<string, ConcreteComponent> | null,
+    fromValue: Record<string, ConcreteComponent> | null,
+    publicThis: ComponentPublicInstance,
+    key: string,
+    asMixin: boolean
+  ) {
+    if (!fromValue) return toValue
+    const ret = toValue ? extend(Object.create(toValue), fromValue) : fromValue
+
+    if (!toValue) {
+      Reflect.setPrototypeOf(ret, { asMixin })
+    }
+
+    return ret
+  }
+})
+;(['methods', 'computed'] as const).forEach(optionName => {
+  defaultMergeStrategies[optionName] = function(
+    toValue?: Record<string, Function>,
+    fromValue?: Record<string, Function>
+  ) {
+    if (!fromValue) return toValue
+    if (!toValue) return fromValue
+    return extend(Object.create(null), toValue, fromValue)
+  }
+})
+
+function normalizeInject(
+  injectOption?: ComponentInjectOptions
+): ObjectInjectOptions | undefined {
+  if (!injectOption) return
+  const normalized = Object.create(null)
+  if (isArray(injectOption)) {
+    for (let i = 0; i < injectOption.length; i++) {
+      normalized[injectOption[i]] = { from: injectOption[i] }
+    }
+    return normalized
+  }
+  return injectOption
+}
+
+function callSyncHook(
+  hooks: Function | Function[],
+  instance: ComponentInternalInstance,
+  type: LifecycleHooks
+) {
+  hooks = isArray(hooks) ? hooks : [hooks]
+  hooks.forEach(hook => {
+    callWithAsyncErrorHandling(hook.bind(instance.proxy!), instance, type)
+  })
+}
+
+function registerHooks(
+  hooks: Function | Function[],
+  hooker: Function,
```

Comment:
Not sure if this is a proper parameter name…

Reply:
hmmm... any suggestions?